cli/azd/pkg/commands/pipeline/azdo_provider.go

```diff
@@ -7,65 +7,528 @@ import (
+		repoDetails.projectName = p.Env.Values[azdo.AzDoEnvironmentProjectName]
+	}
+	if repoDetails.projectId == "" {
+		repoDetails.projectId = p.Env.Values[azdo.AzDoEnvironmentProjectIdName]
+	}
+	if repoDetails.repoName == "" {
+		repoDetails.repoName = p.Env.Values[azdo.AzDoEnvironmentRepoName]
+	}
+	if repoDetails.repoId == "" {
+		repoDetails.repoId = p.Env.Values[azdo.AzDoEnvironmentRepoIdName]
+	}
+	if repoDetails.repoWebUrl == "" {
+		repoDetails.repoWebUrl = p.Env.Values[azdo.AzDoEnvironmentRepoWebUrl]
+	}
+	if repoDetails.remoteUrl == "" {
+		repoDetails.remoteUrl = remoteUrl
+	}
+
 	return &gitRepositoryDetails{
-		owner:    "",
-		repoName: "",
-	}, errors.New("not implemented")
+		owner:    p.repoDetails.orgName,
+		repoName: p.repoDetails.repoName,
+		details:  repoDetails,
+	}, nil
 }
 
 // preventGitPush is nil for Azure DevOps
-func (p *AzdoHubScmProvider) preventGitPush(
+// this method also checks for a credential helper and prompts the user to set a helper to make subsequent pushes easier
+func (p *AzdoScmProvider) preventGitPush(
 	ctx context.Context,
 	gitRepo *gitRepositoryDetails,
 	remoteName string,
 	branchName string,
 	console input.Console) (bool, error) {
-	return false, errors.New("not implemented")
+
+	gitCli := git.NewGitCli(ctx)
+	foundHelper, err := gitCli.CheckConfigCredentialHelper(ctx)
+	if err != nil {
+		return false, err
+	}
+	if !foundHelper {
+		console.Message(ctx, `  
+  A credential helper is not configured for git.
+  This will require you to enter your Azure DevOps PAT when executing a git push.
+  https://aka.ms/azure-dev/git-store
+  `)
+		idx, err := console.Select(ctx, input.ConsoleOptions{
+			Message: "Would you like to enable credential.helper store for this local git repository?",
+			Options: []string{
+				"Yes - set credential.helper = store",
+				"No - do not configure credential.helper",
+			},
+			DefaultValue: "Yes - set credential.helper = store",
+		})
+		if err != nil {
+			return false, fmt.Errorf("prompting for credential helper: %w ", err)
+		}
+		switch idx {
+		// Configure Credential Store
+		case 0:
+			err = gitCli.SetCredentialStore(ctx, p.AzdContext.ProjectDirectory())
+			if err != nil {
+				return false, fmt.Errorf("storing credentials in env: %w ", err)
+			}
+		// Skip
+		case 1:
+			break
+		default:
+			panic(fmt.Sprintf("unexpected selection index %d", idx))
+		}
+	}
+
+	return false, nil
+}
+
+// hook function that fires after a git push
+// allows the provider to perform certain tasks after push including
+// cleanup on the remote url, creating the build policy for PRs and queuing an initial deployment
+func (p *AzdoScmProvider) postGitPush(
+	ctx context.Context,
+	gitRepo *gitRepositoryDetails,
+	remoteName string,
+	branchName string,
+	console input.Console) error {
+
+	gitCli := git.NewGitCli(ctx)
+
+	//Reset remote to original url without PAT
+	err := gitCli.UpdateRemote(ctx, p.AzdContext.ProjectDirectory(), remoteName, p.repoDetails.remoteUrl)
```

Comment:
Swapping out the remote and then swapping it back is a little scary (I am wondering what happens in cases where we swap it and then hit some transient error, do we always swap it back?).  I wonder if we could use the `--repository` argument of `git push` to include the full URI of the remote (with the PAT) so we didn't have to worry about updating remotes.

Reply:
https://github.com/Azure/azure-dev/issues/756